File: app/defi.py
"""Le defi du jour (M14, 5e vague) : le serveur dit quel defi est celui d'aujourd'hui.

Le jeu a six defis (`missions.DEFIS`) : un panneau ou un comptoir, un chrono, une prime — et
la prime ne se paie qu'UNE fois. Le defi du jour est l'un d'eux, **designe par la date**, le
meme pour tout le monde : c'est le serveur qui tient l'horloge, parce qu'un jeu qui tourne dans
le navigateur ne doit pas decider seul du jour qu'on est (un telephone dont l'horloge avance
d'une nuit n'a pas le meme « aujourd'hui » que l'ordinateur).

⚠️ **Il n'y a PAS de classement** : le tableau des scores est parti le 17 sept. 2026, a la
demande de Martin, et le « classement du jour » de la fiche partait avec lui. Ce qui reste du
defi du jour est ce que personne n'a retire : une date, un defi, et une prime qui se gagne
chaque jour.

⚠️ **La graine du jour ne sert qu'a CHOISIR.** La fiche disait que `/api/defi` « donne la
graine du jour » ; or les six defis sont deterministes (ni `B.rng` ni `hash2` : une rampe, un
circuit, une livraison, trois jeux d'adresse a cibles fixes) — un nombre de plus dans la
reponse n'aurait aucun lecteur. On ne livre pas un champ que personne ne lit.

⚠️ **Une rotation, pas un tirage.** `(jour - EPOQUE) % n` : chaque defi revient tous les `n`
jours, jamais deux fois de suite, et le resultat ne depend d'aucun hasard — ni de
`PYTHONHASHSEED` (le piege de `ci-pile-ou-face`) ni du processus : deux workers gunicorn
repondent la meme chose. Ajouter un defi au catalogue le fait entrer dans la rotation ; le
jour du deploiement, le defi du jour peut alors changer — c'est pourquoi le jeu ne paie qu'UNE
prime par jour, quel que soit le defi (`defiDuJour.date`, cote jeu).
"""

from __future__ import annotations

import logging
from datetime import date, datetime, timedelta, timezone, tzinfo
from zoneinfo import ZoneInfo, ZoneInfoNotFoundError

from . import missions

#: Le jour bascule a minuit HEURE DE QUEBEC — pas a minuit UTC, qui tombe a 20 h en ete : un
#: joueur du soir verrait « demain » avant d'avoir fini sa soiree.
FUSEAU = "America/Toronto"

#: Le premier jour de la rotation : le defi du jour est alors le PREMIER du catalogue.
EPOQUE = date(2026, 9, 20)

_journal = logging.getLogger(__name__)


def _fuseau() -> tzinfo:
    """Le fuseau de Quebec — ou, si la base des fuseaux du systeme manque (`tzdata` n'est pas
    une dependance du projet), l'heure normale de l'Est, avec un avertissement.

    ⚠️ Le repli decale la bascule d'une heure en ete (a 1 h au lieu de minuit) : c'est un jour
    qui se change une heure plus tard, pas un defi qui plante ou qui differe d'un worker a
    l'autre. Mieux qu'un 500 sur une route que le titre appelle a chaque chargement.
    """
    try:
        return ZoneInfo(FUSEAU)
    except ZoneInfoNotFoundError:
        _journal.warning("fuseau %s introuvable : repli sur UTC-5 (le jour bascule a 1 h en ete)", FUSEAU)
        return timezone(timedelta(hours=-5), "EST")


def jour_de(maintenant: datetime | None = None) -> date:
    """La date, a Quebec, de cet instant (par defaut : maintenant). `maintenant` porte son fuseau."""
    fuseau = _fuseau()
    if maintenant is None:
        return datetime.now(fuseau).date()
    if maintenant.tzinfo is None:
        raise ValueError("un instant sans fuseau ne dit pas quel jour il est")
    return maintenant.astimezone(fuseau).date()


def rotation() -> list[dict]:
    """Les defis qui tournent : ceux qu'on a des le depart (sans `debloque`).

    ⚠️ Un defi qui se DEBLOQUE (les dix-huit du 23 sept. 2026) n'entre pas dans la rotation :
    le serveur ne connait pas la partie, et il designerait a un joueur neuf un defi qu'il ne
    voit meme pas encore. Et la rotation d'avant ne bouge pas d'un jour."""
    return [d for d in missions.DEFIS if not d.get("debloque")]


def defi_du(jour: date) -> str:
    """Le slug du defi de ce jour : une rotation sur le catalogue, jamais un tirage."""
    catalogue = rotation()
    return catalogue[(jour - EPOQUE).days % len(catalogue)]["slug"]


def aujourdhui(maintenant: datetime | None = None) -> dict:
    """Ce que rend `/api/defi` : la date (celle du serveur) et le slug du defi du jour."""
    jour = jour_de(maintenant)
    return {"date": jour.isoformat(), "defi": defi_du(jour)}

File: static/js/defi.js
/* Bandini — le defi du jour, cote jeu (M14, 5e vague).

   Le jeu a six defis (`B.defs.defis`) : une prime, un chrono, et la prime ne se paie qu'UNE
   fois. Le SERVEUR dit lequel est celui d'aujourd'hui (`GET /api/defi` : `{ date, defi }`) —
   un jeu qui tourne dans le navigateur ne decide pas seul du jour qu'on est —, et le reussir
   paie sa prime UNE FOIS PAR JOUR, meme s'il a deja ete fait.

   ⚠️ IL N'Y A PAS DE CLASSEMENT : le tableau des scores est parti le 17 sept. 2026 (demande de
   Martin) et le « classement du jour » de la fiche partait avec lui. Voir `app/defi.py`.

   ⚠️ SANS RESEAU, IL N'Y A PAS DE DEFI DU JOUR — et le jeu ne s'en apercoit pas. Le defi du
   jour est un bonus, jamais une condition : `duJour()` rend null, le titre ne dit rien, les
   six defis se jouent comme avant. Pas de repli sur l'horloge du telephone : c'est justement
   ce qu'on evite en demandant la date au serveur.

   ⚠️ « DEJA PAYE AUJOURD'HUI » SE COMPARE A LA DATE DU SERVEUR (`B.partie.defiDuJour.date`),
   jamais a un slug ni a l'horloge locale : deployer un defi de plus change la rotation en
   pleine journee, et deux defis du jour le meme jour ne doivent pas payer deux fois.

   ⚠️ CE QUE LE SERVEUR DIT SE VERIFIE : une date qui n'en est pas une, ou un defi que ce
   catalogue ne connait pas (version d'avant, version d'apres), est ignore — comme s'il n'y en
   avait pas. */

const Defi = (function () {
  'use strict';

  //: Au plus une demande toutes les dix minutes : le titre revient souvent, le jour non.
  const REPOS_MS = 10 * 60 * 1000;

  let fenetre = null, url = '/api/defi', courant = null, dernier = 0;
  const abonnes = [];

  function valider(d) {
    if (!d || typeof d !== 'object' || typeof d.defi !== 'string') return null;
    if (typeof d.date !== 'string' || !/^\d{4}-\d{2}-\d{2}$/.test(d.date)) return null;
    // ⚠️ La fiche du catalogue se cherche UNE fois, ici, a la porte : `courant` ne porte jamais un
    // defi que ce catalogue ne connait pas, donc rien d'autre n'a besoin de le reverifier.
    const def = ((B.defs && B.defs.defis) || []).find(function (q) { return q.slug === d.defi; });
    return def ? { date: d.date, slug: d.defi, def: def } : null;
  }

  function prevenir() { for (const f of abonnes) f(courant); }
  function surChangement(f) { abonnes.push(f); }

  /** Demande le defi d'aujourd'hui. ⚠️ Rend TOUJOURS une promesse qui se resout (jamais ne
      rejette) : un reseau coupe, une page d'erreur ou un JSON illisible valent « pas de defi
      du jour », et rien de plus. */
  function actualiser() {
    dernier = Date.now();
    return Promise.resolve().then(function () {
      return fenetre.fetch(url, { cache: 'no-store', credentials: 'omit' });
    }).then(function (r) { return r && r.ok ? r.json() : null; })
      .catch(function () { return null; })
      .then(function (d) {
        courant = valider(d);
        prevenir();
        return duJour();
      });
  }

  function init(w, racine) {
    fenetre = w;
    url = (racine && racine.dataset && racine.dataset.urlDefi) || '/api/defi';
    return actualiser();
  }

  /** Le titre revient, ou la page se rouvre : au plus une demande par `REPOS_MS`. */
  function rafraichir() {
    if (!fenetre || Date.now() - dernier < REPOS_MS) return Promise.resolve(duJour());
    return actualiser();
  }

  /** { date, slug, def } — `def` est la fiche du catalogue —, ou null. */
  function duJour() { return courant; }

  function estDuJour(slug) { return !!courant && courant.slug === slug; }

  /** Cette partie a-t-elle DEJA touche la prime du jour, aujourd'hui (date du serveur) ? */
  function dejaPaye() {
    const f = B.partie && B.partie.defiDuJour;
    return !!courant && !!f && f.date === courant.date;
  }

  /** Vrai si reussir `slug` MAINTENANT paie la prime du jour : c'est le defi d'aujourd'hui,
      et cette partie ne l'a pas encore touchee aujourd'hui. */
  function aPayer(slug) { return estDuJour(slug) && !dejaPaye(); }

  /** La prime du jour est encaissee : la partie s'en souvient, avec la date du serveur. */
  function noterFait(slug, temps) {
    if (!courant) return;
    B.partie.defiDuJour = { date: courant.date, slug: slug, temps: temps };
  }

  return { init, actualiser, rafraichir, surChangement, duJour, estDuJour, dejaPaye, aPayer, noterFait, REPOS_MS };
})();
